server.go

```diff
@@ -287,21 +322,73 @@ func (s *Server) runHandler(handler HandlerFunc, deliveries <-chan amqp.Delivery
 
 		ctx := context.WithValue(context.Background(), CtxQueueName, queueName)
 
-		// TODO: Recover panic in handler. And use delivery.Nack() so amqp will retry.
-		handler(ctx, &rw, delivery)
-		delivery.Ack(false)
+		// Process the request in a separate go routine to not crash the server
+		// on panic and make use of the panic recovery.
+		go func(delivery amqp.Delivery) {
+			pr := processedRequest{
+				replyTo:   delivery.ReplyTo,
+				mandatory: rw.mandatory,
+				immediate: rw.immediate,
+			}
 
-		s.responses <- processedRequest{
-			replyTo:    delivery.ReplyTo,
-			mandatory:  rw.mandatory,
-			immediate:  rw.immediate,
-			publishing: *rw.publishing,
-		}
+			defer s.handlerRecovery(delivery, rw, pr, crashes, &mu)
+
+			handler(ctx, &rw, delivery)
+			delivery.Ack(false)
+
+			pr.publishing = *rw.publishing
+			s.responses <- pr
+		}(delivery)
 	}
 
 	logger.Infof("server: stopped waiting for messages on queue '%s'", queueName)
 }
 
+func (s *Server) handlerRecovery(delivery amqp.Delivery, rw ResponseWriter, pr processedRequest, crashes map[string]float64, mu *sync.RWMutex) {
+	r := recover()
+	if r == nil {
+		return
+	}
+
+	logger.Warnf("server: crashed while handling request")
+
+	// Add new crash to the request recovering from a panic.
+	mu.RLock()
+	crashes[delivery.CorrelationId]++
+	mu.RUnlock()
+
+	switch {
+	case crashes[delivery.CorrelationId] > s.maxRetries:
+		fmt.Fprintf(&rw, "crashed when running handler: %s", r.(string))
+		rw.WriteHeader(ServerCrashedHeader, r.(string))
+
+		pr.publishing = *rw.publishing
+
+		s.responses <- pr
+
+		// Time to remove the requet from crash map, we could not avoid a panic
+		// within the max allowed retry attempts.
+		mu.Lock()
+		delete(crashes, delivery.CorrelationId)
+		mu.Unlock()
+
+		// Nack message and do NOT requeue.
+		delivery.Nack(true, false)
+	default:
+		// Use exponential f(x) = 2^x as backoff but never more than one hour.
+		backoff := math.Pow(2, crashes[delivery.CorrelationId])
```

Comment:
This is dangerous, there could be all sorts of side effects in a handler, it might crash really late. Should we enforce the user to have the handlers idempotent at all times?

Reply:
Yeah that's why the default behaviour is to **not** requeue which I think is a more reasonable behaviour. I thought users would only use `maxRetries` larger than 0 if they have idempotent requests. However, that is only configurable on the whole server and not per handler. :(